platform/mds/mds-web/src/main/java/org/motechproject/mds/web/service/impl/InstanceServiceImpl.java

```diff
@@ -692,61 +714,68 @@ private void setRelationProperty(Object instance, FieldRecord fieldRecord) throw
-                serviceForRelatedClass = DataServiceHelper.getDataService(bundleContext, genericType.getName());
+                argumentType = (Class<?>) ((ParameterizedType) field.getGenericType()).getActualTypeArguments()[0];
             } else if (type.equals(TypeDto.MANY_TO_ONE_RELATIONSHIP) || type.equals(TypeDto.ONE_TO_ONE_RELATIONSHIP)) {
-                serviceForRelatedClass = DataServiceHelper.getDataService(bundleContext, parameterType.getName());
+                argumentType = parameterType;
             }
 
-            value = buildRelatedInstances(serviceForRelatedClass, parameterType, fieldRecord.getValue());
+            serviceForRelatedClass = DataServiceHelper.getDataService(bundleContext, argumentType.getName());
+            Object related = PropertyUtil.safeGetProperty(instance, fieldName);
+
+            value = buildRelatedInstances(serviceForRelatedClass, parameterType, argumentType, fieldRecord.getValue(), related);
         }
 
         Method method = MethodUtils.getAccessibleMethod(instance.getClass(), methodName, parameterType);
         invokeMethod(method, instance, value, methodName, fieldName);
     }
 
-    private Object buildRelatedInstances(MotechDataService service, Class<?> parameterType, Object fieldValue) throws IllegalAccessException, InstantiationException {
-        Object parsedValue = null;
+    private Object buildRelatedInstances(MotechDataService service, Class<?> parameterType, Class<?> argumentType, Object fieldValue, Object related)
+            throws IllegalAccessException, InstantiationException, ClassNotFoundException, NoSuchMethodException, CannotCompileException, NoSuchFieldException {
+        Object parsedValue;
+        RelationshipsUpdate relationshipsUpdate = mapper.convertValue(fieldValue, RelationshipsUpdate.class);
+
+        EntityDto relatedEntity = getEntity(argumentType.getName());
+        List<FieldDto> entityFields = getEntityFields(relatedEntity.getId());
 
        if (Collection.class.isAssignableFrom(parameterType)) {
            Class<?> collectionImplementation = TypeHelper.suggestCollectionImplementation((Class<? extends Collection>) parameterType);
            parsedValue = collectionImplementation == null ? new ArrayList() : collectionImplementation.newInstance();
+           Collection relatedInstances = ((Collection) parsedValue);
 
-           for (Object object : (Collection) fieldValue) {
-               if (isFromUI(object)) {
-                   ((Collection) parsedValue).add(findRelatedObjectById(((Map) object).get(ID_FIELD_NAME), service));
-               } else if (isHistoricalObject(object)) {
-                   String currentVersion = HistoryTrashClassHelper.currentVersion(object.getClass());
-                   ((Collection) parsedValue).add(findRelatedObjectById(PropertyUtil.safeGetProperty(object, currentVersion), service));
+           // Add already related instances, provided any exist
+           if (related != null) {
```

Comment:
1) Change the naming, what's the difference between related and relatedInstances?
2) If we assume related is a collection, do that at the method signature level